libffado/support/tools/ffado-diag.1
.TH FFADO-DIAG 1 27-Mar-2012 "ffado-diag"
.SH NAME
ffado-diag \- print system diagnostic information related to FFADO.
.SH SYNOPSIS
.BI "ffado-diag"
.sp
.SH DESCRIPTION
.B ffado-diag
prints out an extensive collection of diagnostic information about the
computer it is run on.  Information included is the FFADO version number,
the version number of libraries which FFADO depends on, the firewire
interface card, interrupt usage, and so on.  This is useful for developers
to know when giving support via the FFADO mailing lists.  For anything other
than trivial issues, the output of 
.B
ffado-diag
will be one of the first things asked fro when debugging a problem.
.SH "OPTIONS"
.B
ffado-diag
does not presently have any command line options.


ChangeLog
REMARK: Please use the SVN commit messages as replacement for ChangeLog

2007-03-20  Pieter Palmers
    * name change from FreeBoB to FFADO
    * replaced tabs by 4 spaces
    * got rid of end-of-line spaces
    * made all license and copyrights conform
       library becomes LGPL, apps become GPL
       explicitly state LGPL v2.1 and GPL v2 (don't like v3 draft)
       copyrights are 2005-2007 Daniel & Pieter
       except for the MotU stuff (C) Jonathan, Pieter

2006-11-20  Daniel Wagner  <wagi@monom.org>

    * freebob_set_samplerate: return value consitency fix. On success
    it return 0 else -1. Reported by Jamie Gennis jgennis at gmail dot com

    * freebob_streaming_start: Reset the playback ringbuffer before calling 
    freebob_streaming_start. 
    
2006-11-12  Daniel Wagner  <wagi@monom.org>

    * test mixer programm added.

    * basic function block AV/C commands added.
    test program for setting the volume added.

2006-11-11  Daniel Wagner  <wagi@monom.org>

    * src/bebob/bebob_avdevice.cpp (setSamplingFrequencyPlug): copy past
    error fixed. Reported by Jamie Gennis jgennis at gmail dot com

2006-08-27  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version set to 1.4.0
    * maudio device support infrastructure added. Implements
    IAvDevice interface, discover reads the information from
    an xml file. 

2006-08-22  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version bump to 1.3.0
    * freebob_sync allows now to set the sync mode. It does not
    check yet if it really could set the mode.

2006-07-17  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version bump to 1.2.0
    * freebob_sync program added
    * Basic interface added for sync info

2006-07-02  Daniel Wagner  <wagi@monom.org>

    * firmware downloader: check device firmware compatibility

2006-06-27  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version bump to 1.1.0
    * remove bebob_light code
    * downloader various improvements
    * ConfigRom::isAvcDevice() removed. Device probe code added. 
    Each device driver class can check if it supports a device.

2006-06-24  Daniel Wagner  <wagi@monom.org>

    * tests/streaming/Makefile.am: Fixed build problem with missing '/'.

2006-06-21  Daniel Wagner  <wagi@monom.org>

    * freebob-downloader first version added

2006-06-10  Daniel Wagner  <wagi@monom.org>

    * Changed building rules: No static sub libraries
    are build anymore.

2006-06-01  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version bump to 1.0.0

    * Changed all FreeBob to FreeBoB
    * Removed all .cvsignore
    * Added Pieter to AUTHORS
    * Updated NEWS and README (release canditate date added)

2006-05-31  Daniel Wagner  <wagi@monom.org>

    * Free all allocated asprintf buffers after xmlNewChild.
    Compiler warning removed

2006-05-31  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    * src/devicemanager.cpp:
    - fixed some memory leaks
    * src/bebob_light/bebob_light_avdevice.cpp:
    - fixed some memory leaks
    - implemented a test to see if the correct AVC subunits are 
      present
    - fixed a bug that added NULL to the subunit list when an
      unsupported AVC subunit was found.
    * src/libfreebobavc/avc_generic.cpp:
    - fixed memory leak
    * src/libfreebobavc/ieee1394service.cpp:
    - fixed memory leak
    * src/libfreebobavc/avc_extended_plug_info.cpp:
    - fixed memory leak
    * src/libfreebobstreaming/messagebuffer.h:
    - add extern "C"
    * src/libfreebobstreaming/freebob_streaming.c:
    - fixed problem with verbosity setting not being passed on
    - reformatted some error/debug messages
    * configure.ac: version bump to 0.9.4

2006-05-29  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    - implemented a hacked solution to the naming problem when 
      using 2 identical devices. Now the port name is prefixed 
      with "dev#?" with # being the device number (in order of 
      discovery), and the ? being either "c" or "p", meaning 
      "capture" or "playback". Note however that there is no 
      guarantee that device 1 will be the same device next time 
      you start up. This will have to change in 2.0.
    * configure.ac: version bump to 0.9.3

2006-05-26  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    * src/libfreebobstreaming/*: 
    - added a 'verbose' option to disable/enable message output
    - removed some residual messages
    * configure.ac: version bump to 0.9.2

2006-05-24  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    * README:
    - updated readme to reflect recent changes
    * src/libfreebobstreaming/*: 
    - message buffer fix
    * configure.ac: 
    - Made -O2 the default for non-debug builds
    - Version bump to 0.9.1

2006-05-22  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    * */Makefile.am:
    - fixed some problems wrt missing files in SVN and in make dist

    * configure.ac: version bump to 0.9.0 (release candidate 1)

2006-05-12  Pieter Palmers  <pieterpalmers@users.sourceforge.net>
    * src/libfreebobstreaming/*: 
    - Code cleanup
    - Improved SSE code
    - ISO connection parameters are calculated instead of being user
      specified. Note: you need a new jackd backend version due to this.
    
    * configure.ac: Version bumped to 0.8.2

2006-05-06  Daniel Wagner  <wagi@monom.org>

    * This patch fixes two issues when building:
    - libfreebob include files aren't found when build dir != source dir
    - some libraries aren't found when they're not in a system install dir

    Jeremy Kerr

2006-05-04  Pieter Palmers  <pieterpalmers@users.sourceforge.net>

    * configure.ac: Version bumped to 0.8.1

    * Introduced SSE-accellerated packet encoders/decoders

2006-05-04  Daniel Wagner  <wagi@monom.org>

    * configure.ac: Version bumped to 0.8.0

    * Discovering code ported stable 0.2.0 version
    to head and activated for default discovering.

2006-05-04  Daniel Wagner  <wagi@monom.org>


support/tools/ffado-diag.py
#!/usr/bin/python
#

#
# Copyright (C) 2008 Pieter Palmers
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

#
# Test for common FFADO problems
#

import os
import sys
import commands
import re
import logging

from listirqinfo import IRQ,SoftIRQ,IRQInfo
from helpstrings import HelpStrings

## message strings
FFADODIAG_VERSION = "0.1"

welcome_msg = """

FFADO diagnostic utility """ + FFADODIAG_VERSION + """
============================
(C) 2008 Pieter Palmers

"""

help_msg = """
Usage: ffado-diag [verboselevel]

  verboselevel : verbosity level. (optional)

"""

## logging setup
logging.basicConfig()
log = logging.getLogger('diag')

## helper routines

# kernel
def get_kernel_version():
    (exitstatus, outtext) = commands.getstatusoutput('uname -r')
    log.debug("uname -r outputs: %s" % outtext)
    return outtext

def get_kernel_rt_patched():
    print "FIXME: implement test for RT kernel"
    return False

# modules
def check_for_module_loaded(modulename):
    log.info("Checking if module '%s' is loaded... " % modulename)
    f = open('/proc/modules')
    lines = f.readlines()
    f.close()
    for l in lines:
        mod = l.split()[0]
        if mod == modulename:
            log.info(" found")
            return True
    log.info(" not found")
    return False
    
def check_for_module_present(modulename):
    log.info("Checking if module '%s' is present... " % modulename)
    kver = get_kernel_version()
    (exitstatus, outtext) = commands.getstatusoutput("find \"/lib/modules/%s/\" -name '%s.ko' | grep '%s'" % \
                                                     (kver, modulename, modulename) )
    log.debug("find outputs: %s" % outtext)
    if outtext == "":
        log.info(" not found")
        return False
    else:
        log.info(" found")
        return True

def check_1394oldstack_loaded():
    retval = True
    if not check_for_module_loaded('ieee1394'):
        retval = False
    if not check_for_module_loaded('ohci1394'):
        retval = False
    if not check_for_module_loaded('raw1394'):
        retval = False
    return retval

def check_1394oldstack_present():
    retval = True
    if not check_for_module_present('ieee1394'):
        retval = False
    if not check_for_module_present('ohci1394'):
        retval = False
    if not check_for_module_present('raw1394'):
        retval = False
    return retval

def check_1394newstack_loaded():
    retval = True
    if not check_for_module_loaded('fw-core'):
        retval = False
    if not check_for_module_loaded('fw-ohci'):
        retval = False
    return retval

def check_1394newstack_present():
    retval = True
    if not check_for_module_present('fw-core'):
        retval = False
    if not check_for_module_present('fw-ohci'):
        retval = False
    return retval

def check_1394oldstack_devnode_present():
    return os.path.exists('/dev/raw1394')

def check_1394oldstack_devnode_permissions():
    f = open('/dev/raw1394','w')
    if f:
        f.close()
        return True
    else:
        return False

def run_command(cmd):
    (exitstatus, outtext) = commands.getstatusoutput(cmd)
    log.debug("%s outputs: %s" % (cmd, outtext))
    return outtext

# package versions
def get_package_version(name):
    cmd = "pkg-config --modversion %s" % name
    return run_command(cmd)

def get_package_flags(name):
    cmd = "pkg-config --cflags --libs %s" % name
    return run_command(cmd)

def get_command_path(name):
    cmd = "which %s" % name
    return run_command(cmd)

def get_version_first_line(cmd):
    ver = run_command(cmd).split("\n")
    if len(ver) == 0:
        ver = ["None"]
    return ver[0]


def list_host_controllers():
    cmd = "lspci | grep 1394"
    controllers = run_command(cmd).split("\n")
    log.debug("lspci | grep 1394: %s" % controllers)
    for c in controllers:
        tmp = c.split()
        if len(tmp) > 0:
            tmp
            cmd = "lspci -vv -nn -s %s" % tmp[0]
            print run_command(cmd)

## main program
if __name__== '__main__':

    print welcome_msg

    num_args = len(sys.argv)
    if num_args not in [1,2]:
        print help
        sys.exit(0)

    if num_args == 2:
        loglevel = eval(sys.argv[1])
        if loglevel == 1:
            logging.getLogger('diag').setLevel(logging.INFO)
        elif loglevel == 2:
            logging.getLogger('diag').setLevel(logging.DEBUG)

    print "=== CHECK ==="
    print " Base system..."
    
    # check kernel
    kernel_version = get_kernel_version()
    print "  kernel version............ " + str(kernel_version)
    kernel_is_rt_patched = get_kernel_rt_patched()
    print "   RT patched............... " + str(kernel_is_rt_patched)
    
    # check modules
    oldstack_present = check_1394oldstack_present()
    oldstack_loaded = check_1394oldstack_loaded()
    newstack_present = check_1394newstack_present()
    newstack_loaded = check_1394newstack_loaded()
    
    print "  old 1394 stack present.... " + str(oldstack_present)
    print "  old 1394 stack loaded..... " + str(oldstack_loaded)
    print "  new 1394 stack present.... " + str(newstack_present)
    print "  new 1394 stack loaded..... " + str(newstack_loaded)
    
    # check /dev/raw1394 node presence
    devnode_present = check_1394oldstack_devnode_present()
    print "  /dev/raw1394 node present. " + str(devnode_present)
    if devnode_present:
        # check /dev/raw1394 access permissions
        devnode_permissions = check_1394oldstack_devnode_permissions()
        print "  /dev/raw1394 permissions.. " + str(devnode_permissions)
    else:
        devnode_permissions = None

    # check libraries
    print " Prerequisites..."
    print "   gcc................ %s" % get_version_first_line('gcc --version')
    print "   g++................ %s" % get_version_first_line('g++ --version')
    print "   PyQt............... %s" % get_version_first_line('pyuic -version')
    print "   jackd.............. %s" % get_version_first_line('jackd --version')
    print "     path............. %s" % get_command_path('jackd')
    print "     flags............ %s" % get_package_flags("jack")
    print "   libraw1394......... %s" % get_package_version("libraw1394")
    print "     flags............ %s" % get_package_flags("libraw1394")
    print "   libavc1394......... %s" % get_package_version("libavc1394")
    print "     flags............ %s" % get_package_flags("libavc1394")
    print "   libiec61883........ %s" % get_package_version("libiec61883")
    print "     flags............ %s" % get_package_flags("libiec61883")
    print "   libxml++-2.6....... %s" % get_package_version("libxml++-2.6")
    print "     flags............ %s" % get_package_flags("libxml++-2.6")
    print "   dbus-1............. %s" % get_package_version("dbus-1")
    print "     flags............ %s" % get_package_flags("dbus-1")

    # libraw
    
    print " Hardware..."
    # check host controller
    print "   Host controllers:"
    list_host_controllers()
    print "   CPU info:"
    print run_command("cat /proc/cpuinfo")

    print " Configuration..."
    # check RT settings
    
    # check IRQ settings 
    print "  IRQ information"
    info = IRQInfo()

    info.load()
    print str(info)

    print ""
    print "=== REPORT ==="
    
    help = HelpStrings()
    
    # do the interpretation of the tests
    print "FireWire kernel drivers:"
    ## FIXME: what about in-kernel firewire? (i.e. no modules)
    if not oldstack_present:
        help.show('MODULES_OLD_STACK_NOT_INSTALLED')
        sys.exit(-1)
    else:
        if newstack_loaded and oldstack_loaded:
            help.show('MODULES_BOTH_STACKS_LOADED')
            sys.exit(-1)
        elif newstack_loaded:
            help.show('MODULES_NEW_STACK_LOADED')
            sys.exit(-1)
        elif not oldstack_loaded:
            help.show('MODULES_OLD_STACK_NOT_LOADED')
            sys.exit(-1)
        else:
            print "[PASS] Kernel modules present and correctly loaded."

    if not devnode_present:
        help.show('DEVNODE_OLD_STACK_NOT_PRESENT')
        sys.exit(-1)
    else:
        if not devnode_permissions:
            help.show('DEVNODE_OLD_STACK_NO_PERMISSION')
            sys.exit(-1)
        else:
            print "[PASS] /dev/raw1394 node present and accessible."
    
    
    